lib/zstream/protocol.ex

```diff
@@ -1,132 +1,405 @@
 defmodule Zstream.Protocol do
-  @moduledoc false
+  @moduledoc """
+  ZIP file format module, including support for Zip64. Adapted from the
+  gist at
 
+  https://gist.github.com/izelnakri/54361617af44b75895205f2a6c87cf27
+  """
   use Bitwise
 
-  # Specification is available at
-  # https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT
+  @four_byte_max_uint 0xFFFFFFFF
+  @two_byte_max_uint 0xFFFF
 
-  @comment "Created by Zstream"
+  @reader_version 20
+  @writer_version 45
+
+  @default_external_attributes 2_175_008_768
+
+  # bit 3 use data descriptor
+  # bit 11 UTF-8 encoding of filename & comment fields
+  @general_purpose_bit_flag <<0x0008 ||| 0x0800::little-size(16)>>
+
+  @comment_string "Created by Zstream"
+  def comment_string, do: @comment_string
+
+  @force_64 !!System.get_env("ZSTREAM_FORCE_ZIP_64")
+
+  defp storage_mode(options) do
+    {coder, _opts} = Keyword.fetch!(options, :coder)
+    coder.compression_method()
+  end
+
+  def local_file_header(%{
+        c_size: compressed_size,
+        size: uncompressed_size,
+        crc: crc32,
+        options: options,
+        name: filename
+      })
+      when compressed_size > @two_byte_max_uint or uncompressed_size > @four_byte_max_uint or
+             @force_64 do
```

Comment:
could you explain how this is supposed to work? From my understanding, `c_size` and `size` will be always be zero during initialization as we are reading from a stream and we won't know either of these values.